### ISSUE TITLE
visual representation of addresses and hashes … [DISCUSSION]

### ISSUE THREAD

**Issue by andi-apeunit** [aeternity/aepp-components#110]
# Introduction
In this issue I’d like to give you a short introduction to the topic of address and hash visualization and I’d like to ask you to take part in a discussion on how we could implement and style this idea for our æpps and components.

# Purpose
The general purpose of visually representing some address or hash is that we want—and also need—something easily knowable or human-readable when it comes to comparing one thing with another. It’s supercilious to ask our users and developers to compare 32 bytes of base58-hashed information, especially in some risky situation like making a transaction with lots of funds.

# Identicons
We already are using one type of hash visualization in the means of identicons. I _think_ we are using the same identicons as Ethereum-Wallet, which is—if I’m not wrong—a modified version of [blockies](https://github.com/download13/blockies), namely [this one](https://github.com/ethereum/blockies).

GitHub introduced [their identicons](https://github.com/blog/1586-identicons) in August 2013. I still think—after quite some years and also lots of research—they are the most beautiful and functional identicons out there, especially when it comes to different sizes and accessibility—think of all sorts of color vision deficiency.

Another approach towards human-readable address and hash representation is some kind of mnemonic user name. Status, for example, is using a formula consisting of `adjective` `adjective` `animal` which is generated from some script(s) located [here](https://github.com/status-im/status-react/tree/develop/src/status_im/utils/gfycat). There’s also [git mnemonic](https://github.com/ranok/hash-mnemonic) which can be used for creating ”speakable, rememberable translations for git ref hashes”, though I’m not sure if it’s mnemonic.

Lately, I found one [GitHub Pages hosted blog](https://barro.github.io/2018/02/avatars-identicons-and-hash-visualization/) which uses four GitHub-like identicons as article identifier … this is also a very smart concept, as the possible amount of unique identicons is probably smaller than the amount of possible æternity addresses. And it [looks terrific](https://user-images.githubusercontent.com/36665485/40649111-5e263168-6330-11e8-9204-2b3494e18f4d.png)!

# References
[The Ethereum Improvement Proposal repository](https://github.com/ethereum/EIPs/) includes some rather current and active issue concerning [Feedback on “ETH Avatar Standard” Direction](https://github.com/ethereum/EIPs/issues/928) which we maybe should have an eye on.

Good read on [Inclusive Avatars via UX Collective on Medium](https://uxdesign.cc/design-avatars-that-make-sense-and-be-more-inclusive-in-the-process-d4dd6a486ea6), and another good article concerning [Placeholder Avatars](http://morgancarter.com.au/design-solutions/placeholder-avatars) in general.

[IdentiAddress](https://github.com/Cygnusfear/IdentiAddress), a proof-of-concept Chromium/Chrome extension for displaying Ethereum addresses with colored backgrounds. Each background color is defined by hexadecimal color derived from 2–6 characters of the address.

One interesting creative project is [Hashworms](dat://ff176b8c453939142d592e1c2c5eef7c9c69423e1f85bc2560af52e6a8750ab2/index.html) by Raphaël Bastide, which you would maybe want to check out via [Beaker Browser](https://beakerbrowser.com/), or just have a look at [that screenshot](https://d2w9rnfcy7mm78.cloudfront.net/2020120/original_3d66c2ace550cb679e703fa9a8abd165.png).

There’s some more—more or less interesting and/or serious—approaches to identicons which I’d like to share with you and at least write them down somewhere.

– [Avity](https://github.com/Hedronium/Avity)
– [identikon](https://github.com/DarrenN/identikon)

– [monsterID](https://github.com/splitbrain/monsterID)
– [Robohash](https://github.com/e1ven/Robohash)

– [pagan](https://github.com/daboth/pagan)
– [Alien-Avatar Generator](https://github.com/TimPietrusky/alien-avatar-generator)

– [FlatHash](http://flathash.com/)
– [The Iidenticon package](https://markkarpov.com/post/the-identicon-package.html)
– [Vanillicon](https://vanillicon.com/)

# Conclusion
I think—in regards of very lots of use cases and different languages, deficiencies, devices, et cetera—we should offer two approaches for our users and developers.

The first approach would be using either one or more images in the style of GitHub’s identicons. Using more of those images, like GitHub user barro for article identifiers, make up beautiful and unique, simple and compact “words” made from special glyphs. We could even create a typeface for this purpose instead of using image or vector files.

The second approach would be using 2–4 mnemonic words with some specific formula, either `adjective` `adjective` `animal`, or `adjective` `color` `orb`, or … please feel free to propose something, maybe æternity related.

Thanks for reading and happily commenting.

**Comment by andi-apeunit** [comment]
> We have to be careful for many reasons: Most users have encountered mnemonics before in relationship to personal secrets (master private key).

You’re right about some relationship to secrets, though I wouldn’t ultimately generalize that.

Maybe it’s just a matter of wording? Maybe the term “mnemonic phrase” is counter-intuitive, it should be “mnemonic secret” …

Also, though I don’t think the word “mnemonic” is the problem here, we don’t have to call the mnemonic usernames consisting of formula `adjective` `color` `orb`—for example—“mnemonic usernames” … maybe we call them “user identifier” … maybe even “public user identifier” … and in the best case, which we should aim for, “unique public user identifier”.

> We have to make sure that we are not sending mixed signals to users regarding what looks like secret and what looks like sharable information.

As written above. Plus it’s a matter of length, I suppose. Compare a three-word user identifier with a twenty-four-word secret phrase.

Please correct me if I’m wrong, but I think to remember that secret phrases are always displayed in lowercase, by design. If true, maybe we could display unique public user identifiers in uppercase.

> Maybe we scare users not to share information displaying mnemonic user names because they are afraid to loose their private key. Worst case we educate them to feel save to share their private key mnemonic, because they have shared similar looking information several times before.

That’s, both, not what we want. And I’m pretty sure we can get this right with unmistakable wording and design (by which I mean hierarchical structure and visual appearance).

> We have to make sure that collisions are rare […]

Let’s aim for “impossible” …

> […] and that tiny differences in public keys create completely different mnemonics.

Definitely.

> And most important: It has to be hard (complex, expensive) to generate similar looking addresses with the same mnemonic.

Yes. And also we’re using an identicon. So it’s three layers of security, by which I mean complexity and expensiveness to generate three items—address, identicon, public user identifier—almost identical to the items they should look alike.

**Comment by davidyuk** [comment]
Is it possible to use names in AENS instead of identicons? Probably identicons is actual only in the case of signing by a device without connection to the internet (like AirGap or Ledger).

**Comment by andi-apeunit** [comment]
> Is it possible to use names in AENS instead of identicons? Probably identicons is actual only in the case of signing by a device without connection to the internet (like AirGap or Ledger).

In my opinion we should always—if not impossible for reasons—display identicons as graphical identifier plus the unique personal (mnemonic) user identifier or, if registered, the AENS name.

So yes, we should—if there’s one registered—display the AENS name, though instead of the unique personal (mnemonic) user identifier and not instead of the identicon.

**Comment by drhus** [comment]
I've been working on a similar implementation for Hash _(and general purpose large identifier)_ Visualization, but in addition to recognizability, a capacity to describe and transmit both online and via voice channels. I've shared an introduction here https://ethresear.ch/t/wallet-shape-address-human-friendly-visualization-of-wallet-address-momcode/3482 

and you can check out the test-lab >> http://momcode.io/lab
